framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java

```diff
@@ -0,0 +1,372 @@
+ */
+
+/**
+ * This class assists the {@link AnnotatedTypeFactory} by reflectively looking
+ * up the list of qualifier class names in each checker's qual directory, and
+ * then loading and returning it as a set of Annotation Class instances. It can
+ * also look up and load qualifiers from external directories that are passed as
+ * arguments to checkers that have extension capabilities such as the Subtyping
+ * Checker or Units Checker.
+ *
+ * Only annotation classes that have the {@link Target} meta-annotation with the
+ * values of {@link ElementType#TYPE_USE} and/or
+ * {@link ElementType#TYPE_PARAMETER} are loaded. Other type qualifiers must be
+ * manually listed in a checker's annotated type factory by overriding
+ * {@link AnnotatedTypeFactory#createSupportedTypeQualifiers()}.
+ * 
+ *
+ * The methods in this class are called within
+ * {@link AnnotatedTypeFactory#loadTypeQualifiersFromQualDir(boolean, Set)}.
+ *
+ * Type system checker writers may wish to subclass this class if they wish to
+ * implement some custom rules to filter or process loaded annotation classes,
+ * by providing an override implementation of
+ * {@link #createAnnotationMirrorFromClass(Class)}
+ * 
+ * @author jeff luo
+ */
+public class AnnotatedTypeLoader {
+    private BaseTypeChecker checker;
+
+    // For loading from a source package directory
+    private String packageName;
+    private static final String QUAL_PACKAGE_SUFFIX = ".qual";
+
+    // For loading from a Jar file
+    private static final String CLASS_SUFFIX = ".class";
+
+    // For loading from external directories
+    private static final String JAVA_SUFFIX = ".java";
+
+    // Constants
+    private static final char DOT = '.';
+    private static final char SLASH = '/';
+
+    /**
+     * Processing Env used to create an Annotation Builder, which is in turn
+     * used to build the annotation mirror from the loaded Class object
+     */
+    protected ProcessingEnvironment processingEnv;
+
+    // Stores the Resource URL of the qual directory of a Checker class, or AnnotatedTypeFactory class of a particular checker
+    URL resourceURL;
+
+    // Stores a mapping of the Annotation object and the converted AnnotationMirror
+    private Map<Class<? extends Annotation>, AnnotationMirror> loadedAnnotations;
+
+    /** 
+     * constructor for loading type qualifiers defined for a checker
+     * @param checker a base type checker or its subclass
+     */
+    public AnnotatedTypeLoader(BaseTypeChecker checker) {
+        this.checker = checker;
+        processingEnv = checker.getProcessingEnvironment();
+        packageName = checker.getClass().getPackage().getName().replace(SLASH, DOT) + QUAL_PACKAGE_SUFFIX;
+        ProcessPackageNameAndLoadAnnos();
+    }
+
+    /**
+     * Constructor for loading type qualifiers defined for an annotated type factory
+     * @param factory an annotated type factory or its subclass
+     */
+    public AnnotatedTypeLoader(AnnotatedTypeFactory factory) {
```

Comment:
I had it there in the early design and debugging stages of the loader since the trail of thought was that the ATFs directly called the loader within createSupportedTypeQualifiers(). I'll remove this constructor, as it is unnecessary now since a checker and its factory are always within the same package.